corgi/tasks/management/commands/tagcomponent.py

```diff
@@ -0,0 +1,115 @@
+        )
+        parser.add_argument(
+            "--tag-value",
+            dest="tag_value",
+            help="Set tag value",
+        )
+        parser.add_argument(
+            "--operation",
+            dest="operation",
+            help="Set tag operations (get,add,remove)",
+        )
+
+    def handle(self, *args, **options):
+        operation = "get"
+        if options["operation"]:
+            operation = options["operation"]
+        component_name = None
+        if options["component_name"]:
+            component_name = options["component_name"]
+        component_purl = None
+        if options["component_purl"]:
+            component_purl = options["component_purl"]
+        tag_name = None
+        if options["tag_name"]:
+            tag_name = options["tag_name"]
+        tag_value = ""
+        if options["tag_value"]:
+            tag_value = options["tag_value"]
+        self.stdout.write(
+            self.style.SUCCESS(
+                f"operation: {operation}",
+            )
+        )
+        self.stdout.write(
+            self.style.SUCCESS(
+                f"component: {component_name}{component_purl}",
+            )
+        )
+        self.stdout.write(
+            self.style.SUCCESS(
+                f"tag: {tag_name}:{tag_value}",
+            )
+        )
+
+        tag_params = {}
+        if tag_name:
+            tag_params["name"] = tag_name
+        if tag_value:
+            tag_params["value"] = tag_value
+
+        component_params = {}
+        if component_purl:
+            component_params["purl"] = component_purl
+        if component_name:
+            component_params["name"] = component_name
+
+        components = Component.objects.filter(**component_params)
+
+        if operation == "get":
+            component_ids = ComponentTag.objects.filter(**tag_params).values_list(
+                "tagged_model_id", flat=True
+            )
+            for c in Component.objects.filter(uuid__in=component_ids):
+                self.stdout.write(
+                    self.style.SUCCESS(
+                        f"{c.name} {c.nvr} {c.purl}",
+                    )
+                )
+
+        if operation == "remove":
+            if components:
+                for c in components:
+                    ComponentTag.objects.filter(**tag_params, tagged_model_id=c.uuid).delete()
+                self.style.SUCCESS(
+                    f"{operation} tags successful on {components.count()} components.",
+                )
+            else:
+                componenttags = ComponentTag.objects.filter(**tag_params)
+                self.style.SUCCESS(
+                    f"{operation} tags successful on {componenttags.count()} components.",
+                )
+                componenttags.delete()
+
+        if operation == "add":
+            for c in components:
+                tag = ComponentTag.objects.get_or_create(**tag_params, tagged_model_id=c.uuid)
```

Comment:
I just realized our current tag model will lead to duplicate data. We'll create one tag with the same name and value for each component which uses that tag.

Maybe that's expected, but I thought we wanted to support having one tag, like "support": "ooss" for example, that could be shared across multiple components. That way we could update or delete tags in bulk across all components that use them.

We could change the links between tags and components to use a ManyToMany field instead, and then above could become something like:
```suggestion
            for c in components:
                tag = ComponentTag.objects.get_or_create(**tag_params)
                # Directly setting either side of a many-to-many relation is prohibited, must use .set(), .add(), etc.
                tag.tagged_models.add(*components)
```

Reply:
yes, much more efficient representation - nice spot ! we should change this ... maybe as a different issue ? or can raise an issue and bundle fix here.

Reply:
Whatever you'd prefer is fine :smile: I just mentioned here since I saw the code and wasn't sure if it was expected. It's fine to go back and fix later as part of another ticket.